In [5]:
import sys
import pandas as pd
import numpy as np
import scipy
import matplotlib.pyplot as plt
import seaborn as sns

print("Sales Analysis environment is ready!")

Sales Analysis environment is ready!


In [ ]:
df = pd.read_csv("../data/AusApparalSales4thQrt2020.csv")

df.head()

,Date,Time,State,Group,Unit,Sales
0,1-Oct-2020,Morning,WA,Kids,8,20000
1,1-Oct-2020,Morning,WA,Men,8,20000
2,1-Oct-2020,Morning,WA,Women,4,10000
3,1-Oct-2020,Morning,WA,Seniors,15,37500
4,1-Oct-2020,Afternoon,WA,Kids,3,7500


In [ ]:
print("Dataset shape:", df.shape)

print("\nColumn names:")
print(df.columns)

print("Data types:")
print(df.dtypes)

print("\nDataset information:")
df.info()

Dataset shape: (7560, 6)

Column names:
Index(['Date', 'Time', 'State', 'Group', 'Unit', 'Sales'], dtype='str')
Data types:
Date       str
Time       str
State      str
Group      str
Unit     int64
Sales    int64
dtype: object

Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 7560 entries, 0 to 7559
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Date    7560 non-null   str  
 1   Time    7560 non-null   str  
 2   State   7560 non-null   str  
 3   Group   7560 non-null   str  
 4   Unit    7560 non-null   int64
 5   Sales   7560 non-null   int64
dtypes: int64(2), str(4)
memory usage: 354.5 KB


In [12]:
print("Missing values in each column:")
print(df.isna().sum())

print("\nNon-missing values in each column:")
print(df.notna().sum())

Missing values in each column:
Date     0
Time     0
State    0
Group    0
Unit     0
Sales    0
dtype: int64

Non-missing values in each column:
Date     7560
Time     7560
State    7560
Group    7560
Unit     7560
Sales    7560
dtype: int64


### Missing-Value Analysis

The dataset was inspected for missing values using the `isna()` and `notna()` functions.

The `isna().sum()` results showed that every column contained zero missing values. The `notna().sum()` results confirmed that all six columns contained 7,560 valid entries.

Therefore, no rows or columns need to be dropped, and no missing values need to be filled. Retaining the complete dataset prevents unnecessary loss of sales information.

In [13]:
duplicate_count = df.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicate_count}")


Number of duplicate rows: 0


### Duplicate Data Analysis

The dataset was inspected for duplicate records using the `duplicated()` function.

The analysis found **0 duplicate rows**, indicating that each record in the dataset is unique.

Since no duplicate records were identified, no rows were removed during the data cleaning process. The dataset is suitable for further analysis.

In [15]:
print("Unique States:")
print(df["State"].unique())

print(f"\nNumber of unique States: {df['State'].nunique()}")

print("\n" + "=" * 50)

print("\nUnique Groups:")
print(df["Group"].unique())

print(f"\nNumber of Groups: {df['Group'].nunique()}")

print("\n" + "=" * 50)

print("\nUnique Time Values:")
print(df["Time"].unique())

print(f"\nNumber of Time Values: {df['Time'].nunique()}")

Unique States:
<StringArray>
[' WA', ' NT', ' SA', ' VIC', ' QLD', ' NSW', ' TAS']
Length: 7, dtype: str

Number of unique States: 7


Unique Groups:
<StringArray>
[' Kids', ' Men', ' Women', ' Seniors']
Length: 4, dtype: str

Number of Groups: 4


Unique Time Values:
<StringArray>
[' Morning', ' Afternoon', ' Evening']
Length: 3, dtype: str

Number of Time Values: 3


In [18]:
print(repr(df["State"].iloc[0]))
print(repr(df["Group"].iloc[0]))
print(repr(df["Time"].iloc[0]))

'WA'
'Kids'
'Morning'


In [19]:
df["State"] = df["State"].str.strip()
df["Group"] = df["Group"].str.strip()
df["Time"] = df["Time"].str.strip()

### Treatment of Missing Data

The dataset was inspected using the `isna()` and `notna()` functions to identify missing values.

The analysis showed that there were no missing values in any of the six columns. Since all 7,560 records contained complete information, no rows or columns were removed, and no missing values required replacement.

If missing values had been present, the treatment method would have depended on the importance of the affected column. Numeric values such as `Sales` or `Unit` could be replaced with statistical measures such as the mean or median when appropriate, while rows with excessive missing information could be removed to maintain data quality.

In [20]:
# Verify that no leading or trailing whitespace remains

whitespace_check = df[["State", "Group", "Time"]].apply(
    lambda s: s.str.contains(r"^\s|\s$", regex=True).sum()
)

print("Leading/Trailing Whitespace Check:")
print(whitespace_check)

Leading/Trailing Whitespace Check:
State    0
Group    0
Time     0
dtype: int64


In [22]:
print("Verified States:")
print(sorted(df["State"].unique()))

print("\nVerified Groups:")
print(sorted(df["Group"].unique()))

print("\nVerified Time Categories:")
print(sorted(df["Time"].unique()))

Verified States:
['NSW', 'NT', 'QLD', 'SA', 'TAS', 'VIC', 'WA']

Verified Groups:
['Kids', 'Men', 'Seniors', 'Women']

Verified Time Categories:
['Afternoon', 'Evening', 'Morning']


### Data Cleaning Verification

After removing leading and trailing whitespace from the `State`, `Group`, and `Time` columns, the dataset was rechecked.

The verification found zero remaining whitespace issues in all three columns. The categorical values were also reviewed to confirm that only the expected states, demographic groups, and time-of-day categories were present.

The dataset is now clean and ready for normalization and analysis.

## Data Normalization

Min-Max normalization will be applied to the `Unit` and `Sales` columns. This method rescales numeric values to a range between 0 and 1 while preserving the relative differences between observations.

In [23]:
# Apply Min-Max normalization to Unit and Sales

df["Unit_Normalized"] = (
    df["Unit"] - df["Unit"].min()
) / (df["Unit"].max() - df["Unit"].min()
)

df["Sales_Normalized"] = (
    df["Sales"] - df["Sales"].min()
) / (df["Sales"].max() - df["Sales"].min()
)

df.head()


,Date,Time,State,Group,Unit,Sales,Unit_Normalized,Sales_Normalized
0,1-Oct-2020,Morning,WA,Kids,8,20000,0.095238,0.095238
1,1-Oct-2020,Morning,WA,Men,8,20000,0.095238,0.095238
2,1-Oct-2020,Morning,WA,Women,4,10000,0.031746,0.031746
3,1-Oct-2020,Morning,WA,Seniors,15,37500,0.206349,0.206349
4,1-Oct-2020,Afternoon,WA,Kids,3,7500,0.015873,0.015873


In [25]:
print("Unit normalization range:")
print(f"Minimum: {df['Unit_Normalized'].min()}")
print(f"Maximum: {df['Unit_Normalized'].max()}")

print("\nSales normalization range:")
print(f"Minimum: {df['Sales_Normalized'].min()}")
print(f"Maximum: {df['Sales_Normalized'].max()}")

Unit normalization range:
Minimum: 0.0
Maximum: 1.0

Sales normalization range:
Minimum: 0.0
Maximum: 1.0


In [26]:
# Compare original values with normilzed values

df[
    [
        "Unit",
        "Unit_Normalized",
        "Sales",
        "Sales_Normalized"
    ]
].head(10)

,Unit,Unit_Normalized,Sales,Sales_Normalized
0,8,0.095238,20000,0.095238
1,8,0.095238,20000,0.095238
2,4,0.031746,10000,0.031746
3,15,0.206349,37500,0.206349
4,3,0.015873,7500,0.015873
5,10,0.126984,25000,0.126984
6,3,0.015873,7500,0.015873
7,11,0.142857,27500,0.142857
8,15,0.206349,37500,0.206349
9,15,0.206349,37500,0.206349


### Normalization Observation

The normalized `Unit` and `Sales` values are identical for each observation. This occurs because the original `Sales` values maintain a constant proportional relationship with `Unit`.

For the displayed records, sales are calculated as approximately `Unit × 2,500`. Since Min-Max normalization preserves relative position within a column, multiplying every value by the same constant does not change the normalized result.

The normalization was therefore performed correctly.

In [27]:
# Display descriptive statistics for the numeric columns

df[["Unit", "Sales"]].describe()

,Unit,Sales
count,7560.000000,7560.000000
mean,18.005423,45013.558201
std,12.901403,32253.506944
min,2.000000,5000.000000
25%,8.000000,20000.000000
50%,14.000000,35000.000000
75%,26.000000,65000.000000
max,65.000000,162500.000000


In [28]:
print("Unit Statistics")
print(f"Mean: {df['Unit'].mean()}")
print(f"Median: {df['Unit'].median()}")
print(f"Mode: {df['Unit'].mode().tolist()}")
print(f"Standard Deviation: {df['Unit'].std()}")

print("\n" + "=" * 50 + "\n")

print("Sales Statistics")
print(f"Mean: {df['Sales'].mean()}")
print(f"Median: {df['Sales'].median()}")
print(f"Mode: {df['Sales'].mode().tolist()}")
print(f"Standard Deviation: {df['Sales'].std()}")


Unit Statistics
Mean: 18.00542328042328
Median: 14.0
Mode: [9]
Standard Deviation: 12.90140277758643


Sales Statistics
Mean: 45013.5582010582
Median: 35000.0
Mode: [22500]
Standard Deviation: 32253.506943966073


### Descriptive Statistics

The `mean()`, `median()`, `mode()`, and `std()` functions were used to summarize the numeric columns.

For the `Unit` column:
- The average number of units sold was approximately **18**.
- The median number of units sold was **14**.
- The most frequently occurring value (mode) was **9**.
- The standard deviation was approximately **12.90**, indicating noticeable variation in units sold.

For the `Sales` column:
- The average sales amount was approximately **45,014**.
- The median sales amount was **35,000**.
- The most common sales amount was **22,500**.
- The standard deviation was approximately **32,253**, indicating considerable variation in sales values.

In [ ]:
state_sales = df.groupby("State")["Sales"].sum()

print(state_sales)

AttributeError: 'SeriesGroupBy' object has no attribute 'cum'